### ISSUE TITLE
ENH: improving the performance of scalar factor in einsum as modifying the alpha/beta coefficient in BLAS

### ISSUE THREAD

**Issue by Geositta2000** [numpy/numpy#21669]
### Proposed new feature or change:

The motivation for this request is for a small improvement, for matrix/array multiplication/contraction, e.g., `\sum_c 1.0 A[a,b] B[b,c] = C[a,c]` and `\sum_c 0.5 A[a,b] B[b,c] = C[a,c]` , especially when the intermediate dimension is much smaller than other dimensions.

If one puts the prefactors `1.0` and `0.5` into the alpha factor in DGEMM (http://www.netlib.org/lapack/explore-html/d1/d54/group__double__blas__level3_gaeda3cbd99c8fb834a60a6412878226e1.html), the timing will be similar. In `einsum`, it can be quite different. For a small contraction dimension, I may try `einsum` without scalar factor then `np.multiply`, but it may be less efficient than using `DGEMM` once.

Here is an IPython example
```
In [1]: import numpy as np

In [2]: import timeit

In [3]: import time

In [4]: na = nc = 1000

In [5]: nb = 5

In [6]: A = np.random.random((na,nb))

In [7]: B = np.random.random((nb,nc))

In [8]: C = np.einsum('ij,jk->ik', A, B)

In [9]: %timeit np.einsum('ij,jk->ik', A, B)
3.12 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [10]: %timeit np.einsum(',ij,jk->ik', 0.5, A, B)
11.5 ms ± 2.2 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [11]:  %timeit np.multiply(C, 0.5, out = C)
1.16 ms ± 276 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

The step [10] is a factor of 3 more slower than step [9]. The step [11] is about 30% comparing with step [9].

If I use Fortran
```
Program test_dgemm

    Use, Intrinsic :: iso_fortran_env, Only :  wp => real64, li => int64
    integer, parameter :: dp = selected_real_kind(15, 307)
  
   ! Implicit None
  
    Real( dp ), Dimension( :, : ), Allocatable :: a
    Real( dp ), Dimension( :, : ), Allocatable :: b
    Real( dp ), Dimension( :, : ), Allocatable :: c
  
    Integer :: na, nb, nc, m_iter
    Integer :: la, lb, lc
    Integer( li ) :: start, finish, rate
    real(dp) :: sum_time1, sum_time2, alpha, beta
    
    Write( *, * ) 'na, nb, nc?'
    Read( *, * ) na, nb, nc
    Allocate( a ( 1:na, 1:nb ) ) 
    Allocate( b ( 1:nb, 1:nc ) )  
    Allocate( c( 1:na, 1:nc ) )
  
  
  !A[a,b] * B[b,c] = C[a,c]
    Call Random_number( a )
    Call Random_number( b )
    c  = 0.0_dp
 
    alpha = 1.0_dp
    beta = 0.0_dp
    m_iter = 100
  
    write (*,*) 'm_iter average', m_iter 

    sum_time1 = 0.0  
    do m = 1, m_iter     
      Call System_clock( start, rate )
      Call dgemm( 'N', 'N', na, nc, nb, alpha, a , na,  b , nb, beta, c, nc )
      Call System_clock( finish, rate )
      sum_time1 = sum_time1 + Real( finish - start, dp ) / rate  
    end do  
    Write( *, * ) 'Time for dgemm-1', sum_time1 / m_iter
      
  
    c  = 0.0_dp
    alpha = 0.5_dp
    beta = 0.0_dp
    sum_time2 = 0.0  
    do m = 1, m_iter     
      Call System_clock( start, rate )
      Call dgemm( 'N', 'N', na, nc, nb, alpha, a , na, b , nb, beta, c, nc )
      Call System_clock( finish, rate )
      sum_time2 = sum_time2 + Real( finish - start, dp ) / rate  
    end do  
    Write( *, * ) 'Time for dgemm-2', sum_time2 / m_iter  
    Write( *, * ) 'ratio and difference', sum_time2/sum_time1, sum_time2 - sum_time1
  
  End
```

with `gfortran` and `openblas`, I got
```
 na, nb, nc?
1000,5,1000
 m_iter average         100
 Time for dgemm-1   2.0060670000000003E-003
 Time for dgemm-2   2.0285510000000004E-003
 ratio and difference   1.0112080005303912        2.2484000000000115E-003
```
About 1% difference with prefactors `1.0` and `0.5`.

In summary, I don't know if the scalar product in `einsum` would enter alpha/beta coefficient in BLAS. If not, maybe this is a possible addition.

**Comment by Geositta2000** [comment]
By adding `optimize='optimal'`, the scalar product contraction will have a similar timing. `%timeit np.einsum(',ij,jk->ik', 0.5, A, B, optimize='optimal')` `1.77 ms ± 67.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)` and  `%timeit np.einsum('ij,jk->ik', A, B, optimize='optimal')` and `2.03 ms ± 372 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)`. I think this issue can be closed.
